clients/naming_client/host_reactor.go

```diff
@@ -171,3 +176,53 @@ func (hr *HostReactor) asyncUpdateService() {
 		time.Sleep(1 * time.Second)
 	}
 }
+
+// return true when service instance changed ,otherwise return false.
+func isServiceInstanceChanged(oldService, newService *model.Service) bool {
+	oldHostsLen := len(oldService.Hosts)
+	newHostsLen := len(newService.Hosts)
+	if oldHostsLen != newHostsLen {
+		return true
+	}
+	// compare refTime
+	oldRefTime := oldService.LastRefTime
+	newRefTime := newService.LastRefTime
+	if oldRefTime > newRefTime {
+		logger.Warn(fmt.Sprintf("out of date data received, old-t: %v , new-t:  %v", oldRefTime, newRefTime))
+		return false
+	}
+	// sort instance list
+	oldInstance := oldService.Hosts
+	newInstance := make([]model.Instance, len(newService.Hosts))
+	copy(newInstance, newService.Hosts)
+	sortInstance(oldInstance)
```

Comment:
Still effect the cached service.

Reply:
I dont understand.The newService.Hosts wont change after using the copy of it , and i've tried.
can u explain why the cached service will be effected?

Reply:
Sorry.  From the line 95 `hr.serviceInfoMap.Set(cacheKey, *service)`, the cached service has already been replaced. So sort operation for `oldInstance` make no effect.